core/psd/imx6ull/bcb.c

```diff
@@ -63,126 +63,141 @@ int dbbt_block_is_bad(dbbt_t *dbbt, uint32_t block_num)
+	fcb->b0_ecc_type = 0x8; /* FIXME */
+	fcb->b0_ecc_size = 0x0; /* FIXME */
+	fcb->bn_ecc_size = 512;
+	fcb->bn_ecc_type = 0x7;
+	fcb->meta_size = 0x10;
+	fcb->ecc_per_page = 8;
+	fcb->fw1_start = 8 * 64;
+	fcb->fw2_start = 24 * 64;
+	fcb->fw1_size = 0x1;
+	fcb->fw2_size = 0x1;
+	fcb->dbbt_start = 0x100;
+	fcb->bbm_offset = 0x1000; /* FIXME */
+	fcb->bbm_start = 0x0;
+	fcb->bbm_phys_offset = 0x1000;
+	fcb->bch_type = 0x0;
+	fcb->read_latency = 0x0;
+	fcb->preamble_delay = 0x0;
+	fcb->ce_delay = 0x0;
+	fcb->postamble_delay = 0x0;
+	fcb->cmd_add_pause = 0x0;
+	fcb->data_pause = 0x0;
+	fcb->speed = 0x0;
+	fcb->busy_timeout = 0xffff;
+	fcb->bbm_disabled = 1; /* disable badblock marker swapping */
+	fcb->bbm_spare_offset = 0;
+	fcb->disable_bbm_search = 1; /* use only DBBT when loading firmware, TODO: maybe bad block markers instead? */
 
 	fcb->checksum = bcb_checksum(((uint8_t *)fcb) + 4, sizeof(fcb_t) - 4);
 }
 
-int fcb_flash(oid_t oid, fcb_t *fcb_ret)
+int fcb_flash(oid_t oid, const flashsrv_info_t *info)
 {
 	char *sbuf, *tbuf;
 	fcb_t *fcb;
-	int i;
+	unsigned int i, fcb_failed = 0;
 	int err = 0;
+	const unsigned int raw_page_size = info->writesz + info->metasz;
 
-	if ((sbuf = mmap(NULL, FLASH_PAGE_SIZE * 2, PROT_READ | PROT_WRITE, MAP_UNCACHED, OID_NULL, 0)) == MAP_FAILED)
+	sbuf = calloc(4, info->writesz);
+	if (!sbuf)
 		return 1;
 
-	if ((tbuf = mmap(NULL, FLASH_PAGE_SIZE * 2, PROT_READ | PROT_WRITE, MAP_UNCACHED, OID_NULL, 0)) == MAP_FAILED)
-		return 1;
+	tbuf = sbuf + 2 * info->writesz;
 
-	memset(sbuf, 0x0, FLASH_PAGE_SIZE * 2);
 
 	fcb = (fcb_t *)(sbuf);
+	fcb_init(fcb, info);
 
-	fcb_init(fcb);
-	memset(tbuf, 0x0, FLASH_PAGE_SIZE * 2);
+	encode_bch_ecc(sbuf, sizeof(fcb_t), tbuf, raw_page_size, 3);
 
-	encode_bch_ecc(sbuf, sizeof(fcb_t), tbuf,  RAW_FLASH_PAGE_SIZE, 3);
+	/* put 0xff as first metadata byte (6ULL BCB has 32 bytes of metadata) to avoid badblock false positive */
+	tbuf[0] = 0xff;
 
 	for (i = 0; i < BCB_CNT; i++) {
-		err = flashmng_writedev(oid, FCB_START + (i * PAGES_PER_BLOCK * RAW_FLASH_PAGE_SIZE), tbuf, RAW_FLASH_PAGE_SIZE, flashsrv_devctl_writeraw);
+		/* TODO: check for badblock? */
+		err = flashmng_writedev(oid, FCB_START + (i * raw_page_size), tbuf, raw_page_size, flashsrv_devctl_writeraw);
+		/* FIXME: do not error-out, write as many FCB as we can */
 		if (err < 0)
 			break;
-	}
 
-	if (fcb_ret != NULL)
-		memcpy(fcb_ret, sbuf, sizeof(fcb_t));
+		/* verify by reading again */
+		memset(sbuf, 0, 2 * info->writesz);
+		err = flashmng_readraw(oid, FCB_START + (i * raw_page_size), sbuf, raw_page_size);
+
+		for (unsigned int j = 0; j < raw_page_size; ++j) {
+			if (sbuf[j] != tbuf[j]) {
+				printf("FCB FAIL[%4u]: W:0x%02x != R:0x%02x\n", j, tbuf[j], sbuf[j]);
+				fcb_failed += 1;
+				break;
+			}
+		}
+	}
 
-	munmap(sbuf, FLASH_PAGE_SIZE * 2);
-	munmap(tbuf, FLASH_PAGE_SIZE * 2);
+	if (fcb_failed > 0)
+		printf("WARN: %u out of %u FCB are broken\n", fcb_failed, BCB_CNT);
```

Comment:
Misleading message since we break on first error.

Reply:
no, only internal loop is breaking, external will try to write next ones anyway (I'm talking about the case write/read succeeded but the resulting data is different. Even this doesn't mean the FCB is broken as BCH might be able to correct  bit errors via ECC).

This is not a final solution, should probably be improved while testing badblocks in FCB.